Fix prestige confirmations and other story flags

guri256 opened the pull request:
- Fix actions advancing the story even when the action wasn't started
- Fix Imbuement3 story being completed on prestige before it's attempted.
- Fix Imbuement3 and Feast never giving the attempted message
- Fix when guild actions unlock
- Fix mana costs not updating when Mercantilism is leveled
- Fix unlock for Restoration shortcut
- Fix repeated confirmations when prestiging

guri256 committed:
Remove repeated confirmations when selecting prestige abilities
Fix actions advancing the story even when the action wasn't attempted
Fix Imbuement3 story being completed on prestige before it's attempted.
Fix Imbuement3 and Feast never giving the attempted message

Supercopia reviewed (approved):
Very good fixes, well done.

guri256 committed:
There's no reason to unlock Portal if you can't actually do it.  Restoration

guri256 committed:
Don't permanently corrupt the save file if you try to invest NaN amount of gold.

guri256 committed:
When you get more Mercantilism, the Buy Mana actions need to be updated.

guri256 committed:
Fix segmentFinished not being called on the final segment of a capped multi-part action.
I think this bug was only triggering if you required multiple actions to finish the last 2 segments.
This meant the switch(curGodsSegment==63) was never firing off.

guri256 committed:
Fix jungle survivors story unlocking way too early

guri256 committed:
Spatiomancy reduces the cost of "Mana Well" not the number of mana wells.

Supercopia committed:
From Supercopia:
Corrected the requirements for the action stories of ranked multipart actions.

Also aligned the Fight Jungle Monsters to be the same as the other ones. Also, the Thives guild was the only one that was correct, other than the Jungle monsters one.

dmchurch commented on actionList.js:
```diff
@@ -6457,7 +6459,7 @@ Action.OpenPortal = new Action("Open Portal", {
     },
     unlocked() {
-        return getExploreProgress() >= 75;
+        return getExploreProgress() >= 75 && getSkillLevel("Restoration") >= 1000;
```
This shouldn't change. The `unlocked` condition should only test for conditions that are specified in the tooltip as "Unlocked with/at", because those are the only ones that will show up while the action is visible-but-locked. 1000 Restoration is a "Requires" condition, so it has to stay in `canStart`

guri256 replied:
Fixed requirement text for this

dmchurch commented on actionList.js:
```diff
@@ -6426,6 +6426,8 @@ Action.Escape = new Action("Escape", {
         unlockTown(7);
     },
     story(completed) {
+        //FIXME: This will (unfortunately) give the story completion for creating the looping potion, even
+        //if the player didn't, because completing story N will also complete all stories less than N.
```
Yeah, that needs some bigger refactoring than I feel like doing right at this moment and I've been putting it off XD

dmchurch commented on actionList.js:
```diff
@@ -4934,9 +4934,8 @@ Action.Mercantilism = new Action("Mercantilism", {
     },
     finish() {
         handleSkillExp(this.skills);
-        view.requestUpdate("adjustManaCost", "Buy Mana Z1");
-        view.requestUpdate("adjustManaCost", "Buy Mana Z3");
-        view.requestUpdate("adjustManaCost", "Buy Mana Z5");
```
lol, the mana cost for these doesn't change, does it

guri256 replied:
Nope.  I think someone thought "gold cost of bought mana", forgetting that's under GoldCost

dmchurch commented on actionList.js:
```diff
@@ -7063,9 +7065,19 @@ Action.Invest = new Action("Invest", {
+
+        //Looks like something (maybe very high accelerations?) can give you a gold value of NaN.  If so, don't corrupt
+        //the save file
+        if (!isNaN(resources.gold))
```
I think the NaN is a result of the Google Chrome bug, which has since been fixed, but adding some safety checks isn't the worst thing in the world.

I'd prefer for the check to be `isFinite` rather than `!isNaN`, though, since that catches more errors.

guri256 replied:
Sure

guri256 replied:
I do think it's the Chrome bug, but I don't know how long that'll stick around, and if anyone gets NaN saves in their save file, they're pretty screwed.

guri256 replied:
Updated this to isFinite().  Tested it and investment is working.  I didn't test it with NaN, but it should be good.

dmchurch commented on actionList.js:
```diff
@@ -7318,11 +7333,14 @@ Action.ImbueSoul = new MultipartAction("Imbue Soul", {
-            case 4: return buffs["Imbuement"].amt > 499
-                        && buffs["Imbuement2"].amt > 499
-                        && buffs["Imbuement3"].amt > 6;
+            //Protect these with a check for soulInfusionAttempted, so they don't "finish" instantly
+            //on prestige.
+            case 2: return storyFlags.soulInfusionAttempted && buffs["Imbuement3"].amt > 0;
+            case 3: return storyFlags.soulInfusionAttempted && buffs["Imbuement3"].amt > 6;
+            case 4: return storyFlags.soulInfusionAttempted &&
+                        buffs["Imbuement"].amt > 499 &&
+                        buffs["Imbuement2"].amt > 499 &&
+                        buffs["Imbuement3"].amt > 6;
```
Smart, I like it.

dmchurch commented on actions.js:
```diff
@@ -728,7 +728,13 @@ function markActionsComplete(loopCompletedActions) {
     loopCompletedActions.forEach(action => {
         let completed = action.loops - action.loopsLeft;
-        if (action.story !== undefined) action.story(completed);
+        //Test for completed, because all of the actions in .story(completed) assume it was successfully
+        //completed.  Without this we advance the story by putting an action in the list rather than
+        // completing it.  We should really have a hook for failure as well.
+        if (completed > 0 && action.story !== undefined)
+        {
+            action.story(completed);
+        }
```
I'm not thrilled about this fix, but leave it in. This was originally (I'm fairly certain) supposed to provide for being able to giving stories about trying but not succeeding, where the `completed` parameter would be 0, but in practice the only action that makes use of that is the Small Dungeon. Most of the other actions just assume that completed > 0, so this will give the better experience. I'll revisit this once I do the content refactor.

guri256 replied:
That's sort of what I was thinking too.  But none of the checks have the guard of "did it succeed", and I didn't want to change all of the callbacks to add a guard.

dmchurch commented on lang/en-EN/game.xml:
```diff
@@ -1228,7 +1228,7 @@
             <label>Spatiomancy</label>
             <desc>Who knew bending reality to your will could be so useful!</desc>
-            <desc2><![CDATA[Mana Geyser and Mana Well are reduced to the original / (1 + level / 100).<br>
+            <desc2><![CDATA[Mana Geyser and Mana Well action cost are reduced to the original / (1 + level / 100).<br>
```
Please reword:

```
The mana costs for the Mana Geyser and Mana Well actions are reduced...
```

(but yes, good catch, that confused me when I first saw it too)

guri256 replied:
Updated

dmchurch commented on actionList.js:
```diff
@@ -7369,6 +7387,7 @@ Action.ImbueSoul = new MultipartAction("Imbue Soul", {
         return getBuffLevel("Imbuement") > 499 && getBuffLevel("Imbuement2") > 499;
     },
     finish() {
+        setStoryFlag("soulInfusionAttempted")
```
lol

dmchurch commented on actions.js:
```diff
@@ -154,6 +154,10 @@ class Actions {
                 while (curProgress >= loopCost(segment)) {
                     curProgress -= loopCost(segment);
                     // segment finished
+                    if (curAction.segmentFinished) {
+                        curAction.segmentFinished();
+                        partUpdateRequired = true;
+                    }
```
Yeah, this is reasonable. Changes the order of `segmentFinished` and `loopsFinished` getting called, but I doubt that's gonna make any difference at all

guri256 replied:
The reason for this is to move it above the "break;" that made it not called on the last loop

dmchurch commented on prestige.js:
```diff
@@ -125,6 +125,14 @@ function prestigeWithNewValues(nextPrestigeValues, nextPrestigeBuffs) {
 
 function prestigeConfirmation() {
     save();
+
+    if (totals.actions === 0)
+    {
+        //This helps prevent repeated confirmations, and blowing away your backup when picking a second ability.
+        //If you have no actions, then there's nothing worth protecting.
+        return true;
+    }
+
```
Yup, great, makes sense

dmchurch reviewed (changes requested):
Looking pretty good! Just a couple tweaks noted inline, if you don't mind.

guri256 committed:
PR feedback

guri256 committed:
Woops.  This story change was in the wrong branch
